tests/powermon_tests/unit/test_abstract_protocol.py

```diff
@@ -0,0 +1,108 @@
+                "test_responses": [b"(NAK\x73\x73\r", b"(ACK\x39\x20\r",],
+                "regex": "PBT(0[012])$",
+            }
+        }
+        protocol.add_command_definitions(pbt_command_definition_new, "QUERY")
+        
+        #Returns None since it doesn't match the regex
+        pbt_definition = protocol.get_command_with_command_string("PBT")
+        self.assertIsNone(pbt_definition)
+        
+        #Returns since it does match the regex
+        pbt_definition = protocol.get_command_with_command_string("PBT00")
+        self.assertIsNotNone(pbt_definition)
+        
+        self.assertEqual(pbt_definition.code, "PBT")
+        self.assertEqual(pbt_definition.result_type, ResultType.ACK)
+        self.assertEqual(pbt_definition.get_type(), "QUERY")
+        
+    def test_command_defintition_parameters(self):
+        protocol = AbstractProtocol()
+        #This is just here for comparison
+        qed_command_definition_old = {
+            "QED": {
+                "name": "QED",
+                "description": "Daily PV Generated Energy Inquiry",
+                "help": " -- display daily generated energy, format is QEDyyyymmdd",
+                "response_type": ResultType.INDEXED,
+                "response": [
+                    [0, "PV Generated Energy for Day", ResponseType.INT, "Wh", {"icon": "mdi:solar-power", "device-class": "energy", "state_class": "total"},],
+                    [1, "Year", ResponseType.INFO+":cn[3:7]", ""],
+                    [2, "Month", ResponseType.INFO+":calendar.month_name[int(cn[7:9])]", ""],
+                    [3, "Day", ResponseType.INFO+":cn[9:]", ""],
+                ],
+                "test_responses": [
+                    b"(00238800!J\r",
+                ],
+                "regex": "QED(\\d\\d\\d\\d\\d\\d\\d\\d)$",
+            }
+        }
+        qed_command_definition_new = {
+            "QED": {
+                "name": "QED",
+                "description": "Daily PV Generated Energy Inquiry",
+                "help": " -- display daily generated energy, format is QEDyyyymmdd",
+                "result_type": ResultType.INDEXED,
+                "reading_definitions": [
+                    {"index":0, "description":"PV Generated Energy for Day", "reading_type":ReadingType.WATT_HOURS, "response_type":ResponseType.INT, "icon": "mdi:solar-power", "device-class": "energy", "state_class": "total"},
+                ],
+                "parameters": [
+                    {"name":"date", "description":"Date for query", "parameter_type":ParameterType.DATE, "parameter_format":ParameterFormat.YYYYMMDD}
+                ],
+                "test_responses": [
+                    b"(00238800!J\r",
+                ],
+                "regex": "QED(\\d\\d\\d\\d\\d\\d\\d\\d)$",
+            }
+        }
+        
+        protocol.add_command_definitions(qed_command_definition_new, "QUERY")
```

Comment:
Do we still need QUERY / SETTER separation, should these be better defined, more separated, removed entirely...
Always had (in the back of my mind) a more 'understandable' or interchangable command syntax (--getStatus was initial attempt), eg `getBatteryVoltage` etc - still dont really see a way to do that sensibly tho

Reply:
I've been thinking about this as well. I think we can define a device type, like solar controller, then we can define a command and reading that will fulfill the interface. So for each protocol we implement a getBatteryVoltage and then say which command and which reading will return that result.

Reply:
other devices (BMS, Shunt ...) also have batteryVoltage
Maybe more a 'supported commands' for each protocol approach

Reply:
also shoud QUERY be GETTER to be more in line with get_ set_ python way?

Reply:
I think we should move it into the command definition. Maybe just include it in the ResultType. ResultType.COMMAND which will have a reading of an ACK. Then the other types will be getters.

Reply:
yeah not sure we really need the separation / extra 'definition' item
anything we can remove from the definition the better imo